### DISCUSSION TITLE
Matching null object

### DISCUSSION THREAD

**Discussion by ruwentos** [NRules/NRules#291] [Q&A]
I changed a rule a bit to:
```
public class MatchValidResultsByAccountNumbersAndEntryDescriptionsRule : Rule
{
    public override void Define()
    {
        Entry entry = default;
        MatchAllResult doubleResult = default;

        When()
            .Match(() => entry)
            .Query(() => doubleResult, q => q
                .Match<MatchAllResult>(
                    //get all results with the same entry
                    rr => rr.Entry.Id == entry.Id)
                //collect them
                .Collect()
                //but take only those collections with more then one item (same entry in more than one result)
                .Where(c => c.Count() > 1)
                .Select(c => c.FirstOrDefault(r =>
                    //from those, take only that result where the account number is not part of the entry description
                    r.Entry.UnstructuredRemittanceInfo.Contains(r.InvestorAccount.AccountNumber) &&
                    //and the investor account must be active
                    r.InvestorAccount.InvestorAccountState == InvestorAccountStatusEnum.Active)).Where(r=>r != null) 
            );

        Then()
            //create a valid result
            .Do(ctx => AddNewFact(ctx, doubleResult));
    }

    private static void AddNewFact(IContext ctx, MatchAllResult foundResult)
    {
        if (foundResult != null)
            ctx.Insert(new ValidResult(foundResult));
    }
}
```
It throws an **'RhsExpressionError'** **'Object reference not set to an instance of an object.'** in **'Body = {ctx.Insert(new MatchAllResult(bankAccount, entry, investorAccount, cash))}'**. This code belongs to a different rule that ran before and runs just fine without this rule. 
I understand that 'FirstOrDefault' could return null. But also the function 'AddNewFact' did not made it up.
I then solved it by changing the rule to this:
```
public class MatchValidResultsByAccountNumbersAndEntryDescriptionsRule : Rule
{
    public override void Define()
    {
        Entry entry = default;
        MatchAllResult validResult = default;

        When()
            .Match(() => entry)
            .Query(() => validResult, q => q
                .Match<MatchAllResult>(
                    //get all results with the same entry
                    rr => rr.Entry.Id == entry.Id)
                //collect them
                .Collect()
                //but take only those collections with more than one item (same entry in more than one result)
                .Where(c => c.Count() > 1)
                //take only this collection that contains only one result where:
                //-the account is active
                //-account number is part of the entry description
                .Where(c => c.Count(dr =>
                    dr.InvestorAccount.InvestorAccountState == InvestorAccountStatusEnum.Active &&
                    dr.Entry.UnstructuredRemittanceInfo.Contains(dr.InvestorAccount.AccountNumber)) == 1)
                //then take it
                .Select(c => c.First(dr =>
                    dr.Entry.UnstructuredRemittanceInfo.Contains(dr.InvestorAccount.AccountNumber)))
            );

        Then()
            //create a valid result fact
            .Do(ctx => ctx.Insert(new ValidResult(validResult)));
    }
}
```
My questions:
Is there a more easy way to not match a null object?

**Comment by snikolayev** [comment]
@ruwentos first, let me make sure it's clear why the exception is thrown from the RHS of the first rule that inserts the facts. When a new fact is inserted, it propagates through the Rete network, defined by the rules' patterns, and necessary expressions are evaluated. It is during this insertion of the new fact that the exception was thrown.

Regarding the handling of nulls - as you noted, facts cannot be null. The best way to handle it is using a [null object design pattern](https://en.wikipedia.org/wiki/Null_object_pattern). Something along these lines:
```c#
    .Select(c => c.FirstOrDefault(r => condition) ?? Result.Empty) 
```
Where `Result.Empty` would be a static empty result.

I also wanted to ask you for a favor and help me a bit by simplifying your examples and questions to be more generic, minimalistic reproduction scenarios. I'm sure your business domain for which you are trying to write rules is very important to you and is itself very complex. But it's hard for me to figure things out as I have to parse through all this domain complexity, which, honestly, is not related to NRules. It is also less useful to others if your questions are too focused on your domain. I would like the discussions on this GitHub project to be more about the features and capabilities of the engine itself. I hope you understand and take this into account moving forward. Thank you.

**Reply by ruwentos** [reply]
@snikolayev ,
many thanks for your answer. It helps me a lot.

About your request. I understand. I will take note the next time.
